File: Sources/Shared/SemaforoEngine.swift
import Foundation

enum TrafficPhase: String {
    case green, amber, red

    var label: String {
        switch self {
        case .green: return "VERDE"
        case .amber: return "ÁMBAR"
        case .red: return "ROJO"
        }
    }

    var symbol: String {
        switch self {
        case .green: return "🟢"
        case .amber: return "🟡"
        case .red: return "🔴"
        }
    }
}

struct SemaforoSnapshot {
    let phase: TrafficPhase
    let remaining: TimeInterval

    var countdownText: String {
        let seconds = max(0, Int(remaining.rounded(.down)))
        return String(format: "%02d:%02d", seconds / 60, seconds % 60)
    }

    var subtitle: String {
        switch phase {
        case .green: return "tiempo de paso"
        case .amber: return "cambio a rojo"
        case .red: return "hasta próximo verde"
        }
    }
}

enum SemaforoEngine {
    static let greenDuration: TimeInterval = 67
    static let amberDuration: TimeInterval = 3
    static let redDuration: TimeInterval = 590.03
    static let cycleDuration: TimeInterval = greenDuration + amberDuration + redDuration

    static func snapshot(
        at date: Date,
        greenStartTimeMs: Double
    ) -> SemaforoSnapshot {
        var elapsed = date.timeIntervalSince1970 - (greenStartTimeMs / 1000)
        elapsed.formTruncatingRemainder(dividingBy: cycleDuration)

        if elapsed < 0 {
            elapsed += cycleDuration
        }

        if elapsed < greenDuration {
            return .init(
                phase: .green,
                remaining: greenDuration - elapsed
            )
        }

        if elapsed < greenDuration + amberDuration {
            return .init(
                phase: .amber,
                remaining: greenDuration + amberDuration - elapsed
            )
        }

        return .init(
            phase: .red,
            remaining: cycleDuration - elapsed
        )
    }

    /// Devuelve todos los instantes de inicio de verde comprendidos
    /// entre `date` y `date + horizon`.
    static func nextGreenStarts(
        from date: Date,
        greenStartTimeMs: Double,
        horizon: TimeInterval
    ) -> [Date] {
        guard horizon > 0 else { return [] }

        let reference = greenStartTimeMs / 1000
        let now = date.timeIntervalSince1970
        let end = now + horizon

        let cyclesFromReference = (now - reference) / cycleDuration
        var cycleIndex = ceil(cyclesFromReference)

        var firstStart = reference + cycleIndex * cycleDuration

        // Protección frente a errores de coma flotante alrededor del instante exacto.
        if firstStart < now - 0.001 {
            cycleIndex += 1
            firstStart = reference + cycleIndex * cycleDuration
        }

        var result: [Date] = []
        var current = firstStart

        while current <= end + 0.001 {
            result.append(Date(timeIntervalSince1970: current))
            current += cycleDuration
        }

        return result
    }
}

File: android/app/src/main/java/com/txismon/semaforoarnedillo/SemaforoEngine.kt
package com.txismon.semaforoarnedillo

import kotlin.math.ceil

enum class TrafficPhase(val label: String, val symbol: String) {
    GREEN("VERDE", "🟢"),
    AMBER("ÁMBAR", "🟡"),
    RED("ROJO", "🔴");

    val subtitle: String
        get() = when (this) {
            GREEN -> "tiempo de paso"
            AMBER -> "cambio a rojo"
            RED -> "hasta próximo verde"
        }
}

data class SemaforoSnapshot(
    val phase: TrafficPhase,
    val remainingSeconds: Double
) {
    val countdownText: String
        get() {
            val total = remainingSeconds.coerceAtLeast(0.0).toInt()
            return "%02d:%02d".format(total / 60, total % 60)
        }

    val remainingMillis: Long
        get() = (remainingSeconds.coerceAtLeast(0.0) * 1000).toLong()
}

/**
 * Réplica exacta del motor de la app iOS (Sources/Shared/SemaforoEngine.swift).
 * Ciclo fijo: 67 s verde + 3 s ámbar + 590,03 s rojo = 660,03 s.
 * Todo el cálculo es local: no hay red ni backend.
 */
object SemaforoEngine {

    const val GREEN_DURATION = 67.0
    const val AMBER_DURATION = 3.0
    const val RED_DURATION = 590.03
    const val CYCLE_DURATION = GREEN_DURATION + AMBER_DURATION + RED_DURATION

    fun snapshot(nowMillis: Long, greenStartTimeMs: Double): SemaforoSnapshot {
        var elapsed = (nowMillis / 1000.0) - (greenStartTimeMs / 1000.0)
        elapsed %= CYCLE_DURATION

        if (elapsed < 0) {
            elapsed += CYCLE_DURATION
        }

        if (elapsed < GREEN_DURATION) {
            return SemaforoSnapshot(TrafficPhase.GREEN, GREEN_DURATION - elapsed)
        }

        if (elapsed < GREEN_DURATION + AMBER_DURATION) {
            return SemaforoSnapshot(TrafficPhase.AMBER, GREEN_DURATION + AMBER_DURATION - elapsed)
        }

        return SemaforoSnapshot(TrafficPhase.RED, CYCLE_DURATION - elapsed)
    }

    /**
     * Instantes (epoch ms) de inicio de verde entre [fromMillis] y [fromMillis] + [horizonSeconds].
     */
    fun nextGreenStarts(
        fromMillis: Long,
        greenStartTimeMs: Double,
        horizonSeconds: Double
    ): List<Long> {
        if (horizonSeconds <= 0) return emptyList()

        val reference = greenStartTimeMs / 1000.0
        val now = fromMillis / 1000.0
        val end = now + horizonSeconds

        var cycleIndex = ceil((now - reference) / CYCLE_DURATION)
        var firstStart = reference + cycleIndex * CYCLE_DURATION

        // Protección frente a errores de coma flotante en el instante exacto.
        if (firstStart < now - 0.001) {
            cycleIndex += 1
            firstStart = reference + cycleIndex * CYCLE_DURATION
        }

        val result = mutableListOf<Long>()
        var current = firstStart

        while (current <= end + 0.001) {
            result.add((current * 1000).toLong())
            current += CYCLE_DURATION
        }

        return result
    }
}
